### ISSUE TITLE
defaultProject is ignored when using GCE Default Service Account authentication

### ISSUE THREAD

**Issue by chrisleavoy** [grafana/google-bigquery-datasource#101]
The `defaultProject` data source configuration key is ignored when using GCE Default Service Account authentication. This seems like a regression/breaking change from the previous version of this plugin. I'm unable to use a datasource configured with this plugin.  I receive the following error upon testing the data source:

```
403: Access Denied: Project xxx: User does not have bigquery.jobs.create permission in project xxx., accessDenied
```

I believe this issue arises for anyone where BigQuery jobs created by Grafana need to run in a GCP project different from where Grafana is deployed.  I think this is a fairly common situation for billing and BigQuery quota management.

In the doitintl version, administrators could provide the following data source configuration and Grafana would create the bigquery jobs in a project it actually has bigquery access.

```
{"authenticationType":"gce","defaultProject":"yyy","flatRateProject":"yyy","queryPriority":"INTERACTIVE"}
```

For clarity, I have three forks of the BigQuery plugin deployed to a single pod in a Kubernetes cluster with a workload identity bound to a GCE service account:
- v1.0.0 from Grafana 
- v2.0.3 from doitintl
- v3.2.2 (an internal fork)

The same deployment and data source configuration works great in the latter two but receives a 403 using the Grafana version.

**Comment by dprokop** [comment]
@chrisleavoy thanks for opening this, looking into the issue!

**Comment by dprokop** [comment]
@chrisleavoy I have looked into the issue and tried to reproduce it. I have created two GCP projects.

**Project A** contains:
- a VM running Grafana, using `Compute Engine default service account`
- a BigQuery instance (just for testing purposes)

**Project B** contains:
- a BigQuery instance. I've added  `Compute Engine default service account` with BigQuery Job User and BigQuery Data Viewer roles via IAM settings

I can query both Project A and Project B BigQuery projects. 

We currently don't provide code completion for multiple projects - we will try to introduce this soon. Moreover, the visual query builder doesn't allow project selection when in GCE scenario - I think we will be able to add that soon.

I don't know how the above helps your case, but if you could either confirm your setup is similart to the above, or provide some more details on your setup then I hope we will be able to help you more and hopefully solve the problem very soon.

**Comment by chrisleavoy** [comment]
Hi @dprokop, thanks for looking into this issue.  

In your attempt to reproduce, Project B would be the default project, not A. Moreover, to see the 403 error, the local Project A permissions for BigQuery would need to be removed.

Our setup is slightly different:  Project A only has Grafana running – there is no local BigQuery dataset or permissions to create BigQuery jobs there. Project B has a BigQuery dataset and all the necessary permissions for Project A's service account.

The way I understand it, BigQuery creates lambda jobs when running queries against a dataset. Google uses these jobs for usage billing. We centralize where users must instantiate jobs for BigQuery to keep cost controls and quotas configuration sane in our environment.

I just tested the 1.0.1 patch, and the behaviour changed slightly.  It still doesn't respect the `defaultProject`.  When configuring the data source, the `Save & test` action still shows Grafana uses the local project to create BigQuery jobs.  The Google API rightfully returns a 403 error.  As it worked in the last fork and other Google plugins like StackDriver, we expect the `defaultProject` to override where Grafana creates BigQuery jobs.

**Comment by dprokop** [comment]
@chrisleavoy thanks for the detailed description. I think I have managed to reproduce that exact environment.

**Project A**: Grafana instance. _Default service account_ with the default Editor role.
**Project B**: Jobs runner. Added Project A's _default service account_ with `BigQuery Job User` role.
**Project C**: BigQuery instance. Added Project's A _default service account_ with `BigQuery Data Viewer` role.


As far as I can see the implementation problem is, as you described, not using the default project configured when creating the jobs. Indeed, no matter the default project config, it was never used when creating a BigQuery client. I'm bringing this in with this PR: https://github.com/grafana/google-bigquery-datasource/pull/103. I hope to release `1.0.2` later today, so you could verify it works in your setup.

Unfortunately, I can see one limitation that we need to solve soon for the described GCE setup to be user-friendly. There may be several BQ projects that one wants to query. We need to introduce [Resource Managere API](https://cloud.google.com/resource-manager) integration, similar to the Google Cloud Monitoring data source, so we can allow projects selection when creating the query. As of now, it's only the GCE default project(not the one introduced by #103) that gets code editor support and visual query builder datasets and tables selection. With #103 in place one will be able to create queries using Code editor only (visual query builder doesn't allow project selection), but the code completion will be broken.

@chrisleavoy Let me know if the 1.0.2 makes the thing work for you, and in the meantime I'll land on a more detailed description of what needs to be done in the query builder, to support code completion and visual query builder given described setup.

**Comment by chrisleavoy** [comment]
@dprokop, I just tried 1.0.2 and GCE auth is working now, thank you!

Regarding the Explorer view.  I like the changes, but yes, being able to select the project would be nice.  Otherwise, we would have to create a few dozen BigQuery data sources for all the different projects we have internally.

I'll do some more testing today. I saw a bunch of 500 internal errors after switching a few panels to the new plugin. I will open a new issue after I discover the trigger.
